eni: Add garbage collector for leaked ENIs

gandro opened the pull request:
This PR adds a new garbage collector for leaked ENIs. ENIs might be
leaked if the operator is unable to perform an ENI delete operation, for
example because the EC2 API is (temporarily) unavailable or the operator
crashed before it was able to issue the delete operation.

The GC runs in a regular interval and removes any ENIs with a predefined
set of tags. The tags are added to newly created ENIs, meaning this GC
does not remove any ENIs created by older versions of Cilium. This is
done to avoid removing any ENIs which were never managed by Cilium.

The tags used for ENI creation and deletion, as well as the GC interval,
can both be configured using operator flags. If the GC interval is set
to zero, the ENI GC is disabled.

**Review per commit**

gandro committed:
aws: Rename InstancesFilter to TagsFilter

The function is generic for all tags-based filters.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
ec2/mock: Fix semantics of DeleteNetworkInterface

The DeleteNetworkInterface method does not allow deleting already
attached interfaces. This commit changes the mock implementation
behavior to match the behavior of the actual EC2 API.

Some of the existing unit tests relied on the old behavior. These tests
are updated to manually detach the mock ENI where necessary.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
aws/ec2: Add GetDetachedNetworkInterfaces method

This new method returns all detatched (aka. available) network
interfaces matching a given tag filter. This will be used to clean up
those ENIs in case of resource leaks.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
eni: Add ENI garbage collector

This commit adds a new garbage collector for leaked ENIs. ENIs might be
leaked if the operator is unable to perform a delete operation, for
example because the EC2 API is (temporarily) unavailable or the operator
crashed before it was able to issue the delete operation.

The GC runs in a regular interval and removes any ENIs with a predefined
set of tags. The tags are added to newly created ENIs, meaning this GC
does not remove any ENIs created by older versions of Cilium. This is
done to avoid removing any ENIs which were never managed by Cilium.

The tags used for ENI creation and deletion, as well as the GC interval,
can both be configured using operator flags. If the GC interval is set
to zero, the ENI GC is disabled.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
helm: Add flags to tweak eni.gc{Interval,Tags}

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

joestringer reviewed (approved):
The docs changes look trivially correct. Please re-request my review if you'd like me to take a closer look at the logic, otherwise I will leave it to the corresponding codeowners.

pippolo84 reviewed (approved):
LGTM :+1:

squeed commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -588,6 +588,13 @@ eni:
+  gcInterval: ""
+  # -- Additional tags attached to ENIs created by Cilium.
+  # Dangling ENIs with this tag will be garbage collected
+  # @default -- `{"io.cilium/cilium-managed":"true}`
```
Is there ever a chance that we could accidentally select ENIs owned by a different cluster? Should we tag this with a cluster name?

gandro replied:
Yes, we will very likely collect ENIs owned by different clusters. I don't consider this a huge problem (the mechanism is explicitly designed to not need synchronization). But if users want/need to turn off GC for a particular cluster, they will have to either use per-cluster tags here, or set a global subnet tag filter unique to each cluster.

Unfortunately, I expect most users to not have a clustername set. `cilium install` does set one automatically, but most advanced users will use Helm and thus will have an empty clustername. But maybe we should still automatically add a cluster tag here _if_ the clustername is set and recommend it in the docs for ENI IPAM based installations.

squeed replied:
Given that this only collects detached ENIs (correct me if I'm wrong), I assume it's not a big deal if we GC ENIs not "owned" by us, right? If so, this discussion is mostly academic.

But could we theoretically step on an ENI in the creation process?

christarazi replied:
@squeed I believe this is handled by the GC logic waiting for the next cycle to GC ENIs. Basically, it's like a mark and sweep GC. First round we collect all the ENIs (mark) and the next round we delete (sweep). Please correct me if I'm wrong @gandro.

gandro replied:
@christarazi Yes, that's correct. Given an offline discussion with @squeed I will try to add a best-effort clustername tag as well, we might even be able to extract the EKS clustername from the EC2 API where available.

While generally speaking I don't think it is a problem if we GC ENIs from other clusters, if GC misbehaves and the user wants to turn it off or reconfigure, then cross-cluster GC will become an issue (since the settings in a different cluster will affect the GC in the local cluster and vice-versa).

gandro replied:
@squeed @christarazi I have updated the PR to auto-detect the clustername and tag ENIs with the clustername if available. Please take a look.

sayboras reviewed (approved):
I don't have other comment :100:

gandro committed:
ec2: Add DetectEKSClusterName helper

This helper attempts to auto-detect the EKS cluster name from the
instance tag of the instance where cilium-operator is running on.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
docs: Add short section about ENI GC

Namely discuss the implications of unnamed clusters and how to override
the default behavior.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
